db/dbformat.h

```diff
@@ -504,6 +516,61 @@ class IterKey {
     key_size_ = total_size;
   }
 
+  // A version of `TrimAppend` assuming the last bytes of length `ts_sz` in the
+  // user key part of `key_` is not counted towards shared bytes. And the
+  // decoded key needed a min timestamp of length `ts_sz` pad to the user key.
+  void TrimAppendWithTimestamp(const size_t shared_len,
+                               const char* non_shared_data,
+                               const size_t non_shared_len,
+                               const size_t ts_sz) {
+    std::string kTsMin(ts_sz, static_cast<unsigned char>(0));
+    std::string key_with_ts;
+    std::vector<Slice> key_parts_with_ts;
+    if (IsUserKey()) {
+      key_parts_with_ts.reserve(3);
+      key_parts_with_ts.emplace_back(key_, shared_len);
+      key_parts_with_ts.emplace_back(non_shared_data, non_shared_len);
+      key_parts_with_ts.emplace_back(kTsMin);
+    } else {
+      assert(shared_len + non_shared_len >= kNumInternalBytes);
+      // Invaraint: shared_user_key_len + shared_internal_bytes_len = shared_len
+      const size_t user_key_len = key_size_ - kNumInternalBytes;
+      const size_t sharable_user_key_len = user_key_len - ts_sz;
+      const size_t shared_user_key_len =
+          std::min(shared_len, sharable_user_key_len);
+      const size_t shared_internal_bytes_len =
+          shared_len > sharable_user_key_len
+              ? shared_len - sharable_user_key_len
+              : 0;
```

Comment:
This could probably be simply `shared_len - shared_user_key_len`

Reply:
Thanks for the suggestion! I tried this one and it caused some error because when `shared_len` is smaller than `sharable_user_key_len`, that's when all shared bytes come from the user key part, this statement could be a negative number, and the `size_t` type would wrap it around.